[beta] Backport accepted PRs to 1.21

dtolnay opened the pull request:
Backport of:

- ~don't suggest placing `use` statements into expanded code https://github.com/rust-lang/rust/pull/44215~
- stabilize tcpstream_connect_timeout https://github.com/rust-lang/rust/pull/44563
- stabilized iterator_for_each https://github.com/rust-lang/rust/pull/44567
- travis: Move sccache to the us-west-1 region https://github.com/rust-lang/rust/pull/44574
- stabilized ord_max_min https://github.com/rust-lang/rust/pull/44593
- stabilized compiler_fences https://github.com/rust-lang/rust/pull/44595
- ci: Upload/download from a new S3 bucket https://github.com/rust-lang/rust/pull/44617
- stabilized needs_drop https://github.com/rust-lang/rust/pull/44639
- Stabilized vec_splice and modified splice tracking issue https://github.com/rust-lang/rust/pull/44640
- Backport libs stabilizations to 1.21 beta https://github.com/rust-lang/rust/pull/44824

budziq committed:
stabilize tcpstream_connect_timeout (closes #43079)

alexcrichton committed:
travis: Move sccache to the us-west-1 region

Most of the other rust-lang buckets are in us-west-1 and I think the original
bucket was just accidentally created in the us-east-1 region. Let's consolidate
by moving it to the same location as the rest of our buckets.

budziq committed:
stabilized ord_max_min (fixes #25663)

budziq committed:
stabilized compiler_fences (fixes #41091)

budziq committed:
Added example to `compiler_fence` docs taken from unstable-book

budziq committed:
stabilized iterator_for_each (closes #42986)

updated clippy and rls as it uses the iterator_for_each

alexcrichton committed:
ci: Upload/download from a new S3 bucket

Moving buckets from us-east-1 to us-west-1 because us-west-1 is where
rust-central-station itself runs and in general is where we have all our other
buckets.

budziq committed:
Added more text from unstable-book to `compiler_fence` docs

budziq committed:
stabilized needs_drop (fixes #41890)

budziq committed:
stabilized vec_splice (fixes #32310)

budziq committed:
Updated tracking issue for String::splice and its unstable-book entry

dtolnay committed:
Backport libs stabilizations to 1.21 beta

This includes the following stabilizations:

- tcpstream_connect_timeout https://github.com/rust-lang/rust/pull/44563
- iterator_for_each https://github.com/rust-lang/rust/pull/44567
- ord_max_min https://github.com/rust-lang/rust/pull/44593
- compiler_fences https://github.com/rust-lang/rust/pull/44595
- needs_drop https://github.com/rust-lang/rust/pull/44639
- vec_splice https://github.com/rust-lang/rust/pull/44640

kennytm commented on src/liballoc/string.rs:
````diff
@@ -1420,7 +1420,7 @@ impl String {
     /// assert_eq!(s, "Α is capital alpha; β is beta");
     /// ```
-    #[unstable(feature = "splice", reason = "recently added", issue = "32310")]
+    #[unstable(feature = "splice", reason = "recently added", issue = "44643")]
````
Please rename the feature.

```
[00:04:19] tidy error: /checkout/src/liballoc/string.rs:2253: mismatches to previous in: ["tracking issue"]
```

alexcrichton reviewed (approved):
r=me wit hthe tidy error fixed,

thanks @dtolnay!

dtolnay committed:
Use the new String::splice tracking issue for string::Splice

I backported the stabilization of Vec::splice and the move of
String::splice to the new tracking issue, but not the new signature of
String::splice in 1.22. This combination requires updating the tracking
issue on some code that no longer exists in nightly.

dtolnay commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit a03e765 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton a03e7658c6f95f79b1b040a52bad915a3174b073 -->

alexcrichton commented:
@bors: r-

backport fails to compile?

dtolnay commented:
I removed the backport of #44215 from this PR, will leave that one to someone else.

<details>
<summary>(expand) Here is the merge I botched.</summary>

```rust
            match item.node {
                ItemKind::Use(..) => {
                    // don't suggest placing a use before the prelude
                    // import or other generated ones
<<<<<<< HEAD
                    if item.span == DUMMY_SP {
                        let mut span = item.span;
                        span.hi = span.lo;
                        self.span = Some(span);
||||||| parent of 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                    if item.span == DUMMY_SP {
                        self.span = Some(item.span.with_hi(item.span.lo()));
=======
                    if item.span.ctxt().outer().expn_info().is_none() {
                        self.span = Some(item.span.with_hi(item.span.lo()));
>>>>>>> 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                        self.found_use = true;
                        return;
                    }
                },
                // don't place use before extern crate
                ItemKind::ExternCrate(_) => {}
                // but place them before the first other item
                _ => if self.span.map_or(true, |span| item.span < span ) {
<<<<<<< HEAD
                    let mut span = item.span;
                    span.hi = span.lo;
                    self.span = Some(span);
||||||| parent of 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                    self.span = Some(item.span.with_hi(item.span.lo()));
=======
                    if item.span.ctxt().outer().expn_info().is_none() {
                        // don't insert between attributes and an item
                        if item.attrs.is_empty() {
                            self.span = Some(item.span.with_hi(item.span.lo()));
                        } else {
                            // find the first attribute on the item
                            for attr in &item.attrs {
                                if self.span.map_or(true, |span| attr.span < span) {
                                    self.span = Some(attr.span.with_hi(attr.span.lo()));
                                }
                            }
                        }
                    }
>>>>>>> 74748b11bb3... WIP: don't suggest placing `use` statements into expanded code
                },
```
</details>

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit f38d353 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton f38d35386d29e76c6007d47c986001ef7f5877ff -->

alexcrichton commented:
@bors: p=1 (backport)

bors commented:
:hourglass: Testing commit f38d35386d29e76c6007d47c986001ef7f5877ff with merge 3c3cd3f68233aeeccf48d377c2b5f25f72983583...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4801), [status-travis](https://travis-ci.org/rust-lang/rust/builds/279989278?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 3c3cd3f68233aeeccf48d377c2b5f25f72983583 to beta...
